dataset/src/main/scala/frameless/ops/RelationalGroupsOps.scala

```diff
@@ -0,0 +1,212 @@
+package frameless
+package ops
+
+import org.apache.spark.sql.catalyst.analysis.UnresolvedAlias
+import org.apache.spark.sql.catalyst.plans.logical.{MapGroups, Project}
+import org.apache.spark.sql.{Column, Dataset, FramelessInternals, RelationalGroupedDataset}
+import shapeless.ops.hlist.{Mapped, Prepend, ToTraversable, Tupler}
+import shapeless.{::, HList, HNil, ProductArgs}
+
+/**
+  * @param groupingFunc functions used to group elements, can be cube or rollup
+  * @tparam T the original `TypedDataset's` type T
+  * @tparam TK all columns chosen for aggregation
+  * @tparam K individual columns' types as HList
+  * @tparam KT individual columns' types as Tuple
+  */
+private[ops] abstract class RelationalGroupsOps[T, TK <: HList, K <: HList, KT]
+  (self: TypedDataset[T], groupedBy: TK, groupingFunc: (Dataset[T], Seq[Column]) => RelationalGroupedDataset)
+  (implicit
+   i0: ColumnTypes.Aux[T, TK, K],
+   i1: ToTraversable.Aux[TK, List, UntypedExpression[T]],
+   i2: Tupler.Aux[K, KT]
+  ) {
+  object agg extends ProductArgs {
+    /**
+      * @param i3 shares individual columns' types
```

Comment:
I feel that comments on these kind of implicits do not belong to the scaladoc but should go alongside the code instead. Something like this:

```scala
def applyProduct[TC <: HList, C <: HList, OptK <: HList, Out0 <: HList, Out1]
(columns: TC)
(implicit
  i3: AggregateTypes.Aux[T, TC, C], // shares individual columns' types
  i4: Mapped.Aux[K, Option, OptK],  // maps all types in HList to Option
  i5: Prepend.Aux[OptK, C, Out0],   // concatenates two HLists
  i6: Tupler.Aux[Out0, Out1],       // converts HList to Tuple
  i7: TypedEncoder[Out1],           // proof that there is `TypedEncoder` for the output type
  i8: ToTraversable.Aux[TC, List, UntypedExpression[T]] // allows converting thi HList to ordinary List
): TypedDataset[Out1] = {
```

Reply:
I think you're right, I wanted to be consistent with `SmartProject` but in case of `RelationalGroupsOps` with two implicit lists it's hard to read it like that